### PR TITLE
Create Budget Web application

### PR DESCRIPTION
**Pull request by tjay1760**
# In this pull request I created a budget App with the following specifications: 

### Splash screen

A simple page with the name of my app and links to the sign up and log in pages.

### Sign up and log in pages

The user is able to register in the app with full name, email and password 
The user can log into the app using email and password.
If the user is not logged in, they can't access pages that require the user to be logged in 

### Home page (categories page)

When the user logs in, they are presented with the categories page.
For each category, the user can see their name, icon and the total amount of all the transactions that belongs to that category.
When the user clicks (or taps) on a category item, the application navigates to the transactions page for that category.
There is a button "add a new category" at the bottom that brings the user to the page to create a new category.

### Transactions page

For a given category, the list of transactions is presented, ordered by the most recent.
At the top of the page the user could see the total amount for the category (sum of all of the amounts of the transactions in that category).
There is a button "add a new transaction" at the bottom that brings the user to the page to create a new transaction.
When the user clicks on the "Back" button (<), the user navigates to the home page.

### "Add a new category" page

The user fills out a form to create a new category, indicating their name and icon (both mandatory).
The user clicks (or taps) the "Save" button to create the new category, and is taken to the home page on success.
When the user clicks on the "Back" button (<), the user navigates to the home page.

### "Add a new transaction" page

The user fills out a form to create a new transaction with:
name 
amount 
categories 
The user clicks (or taps) the "Save" button to create the new transaction, and is taken to the transactions page for that category.
When the user clicks on the "Back" button (<), the user navigates to the transactions page for that category.
I also created unit and integration tests for all the most important components of the application.

### Other technical features

I used Postgres as your database.
I used devise for authentication.
I validated all user input to make sure that anyone with bad intentions cannot compromise the app.
The project  deployed and accessible online here: https://budgetapp-g2pb.onrender.com.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,24 +1,169 @@
...
+
+## ⭐️ Show your support <a name="support"></a>
+
+Please support this project and be a part of something meaningful and impactful. Your contribution can make a significant difference and help us achieve our goals faster.
+
+<p align="right">(<a href="#readme-top">back to top</a>)</p>
+
+<!-- ACKNOWLEDGEMENTS -->
+
+## 🙏 Acknowledgments <a name="acknowledgements"></a>
```
Comment (fobadara) [inline README.md:143]: [REQUIRED] Kindly give proper credit to Gregoire Vella the owner of the design according to  the [Creative Commons License](https://creativecommons.org/licenses/by-nc/4.0/)

File: app/views/main/index.html.erb
Comment (fobadara) [inline app/views/main/index.html.erb:1]: [REQUIRED] Kindly ensure that your design matches the template in terms of colour typography and layout. Kindly centre the app name and ensure that the login nad signup buttons are at the bottom.

image

[OPTIONAL] I noticed that you considered another page as the splash page but based on the instructions the splash page is supposed to contain your app name and the login and signup buttons.

File: app/views/devise/registrations/new.html.erb
Comment (fobadara) [inline app/views/devise/registrations/new.html.erb:1]: [REQUIRED] Kindly ensure that your signup page looks like the template in the image below. You should have the `Register` button in the center, your submit button should be the `Next` button on the top right of the template, Your inputs should fill the width of the screen and not have spaces

image

[REQUIRED] Also look at the image below for other required changes:
image

File: app/views/devise/sessions/new.html.erb
Comment (fobadara) [inline app/views/devise/sessions/new.html.erb:1]: [REQUIRED] Kindly ensure that your design matches the template in terms of colour, typography and layout. Kindly look at the image below and make adjustments as specified:
image

File: app/views/categories/index.html.erb [hunk 2]
```diff
@@ -0,0 +1,12 @@
+<%= stylesheet_link_tag 'categories', media: 'all', 'data-turbolinks-track': 'reload' %>
+
+<div class ="categories">
+  <%= link_to "New category", new_category_path, class: "link_to newcat" %>
```
Comment (fobadara) [inline app/views/categories/index.html.erb]: [REQUIRED] Kindly move your `new category` button to the bottom of the page. It should be replacing the scan button.
image

File: app/views/categories/index.html.erb [hunk 3]
```diff
@@ -0,0 +1,12 @@
...
+  <% if @categories.empty? %>
+      <p>No categories available</p>
+        <% else %>
+    <% @categories.each do |category| %>
+          <%= link_to render(category), category_path(category), class: "link_to cats" %>
+    <% end %>
```
Comment (fobadara) [inline app/views/categories/index.html.erb]: [REQUIRED] Please ensure that your design matches the template. Kindly look at the image below:
image

File: app/views/expenses/index.html.erb
Comment (fobadara) [inline app/views/expenses/index.html.erb:1]: [REQUIRED] Kindly ensure that the design for your transaction page looks more like the transaction or details page in the Behance template in terms of colour, typography and especially layout. The general layout doesn't look like any of them

[REQUIRED] When I try to create an expense without selecting a category I get the error in the image below. Kindly fix this problem and ensure that the user can't submit the form without selecting at least one category.

image

File: config/routes.rb [hunk 4]
```diff
@@ -1,6 +1,12 @@
 Rails.application.routes.draw do
+  devise_for :users
   # Define your application routes per the DSL in https://guides.rubyonrails.org/routing.html
 
   # Defines the root path route ("/")
-  # root "articles#index"
+   root "splash#index"
+  resources :main, only: %i[index]
+  resources :users
+  resources :categories, only: %i[index new create update show destroy]  do
```
Comment (fobadara) [inline config/routes.rb]: [REQUIRED] You include an update under the created resources even though you have no action for it in your controller. Kindly ensure that you are creating only resources that are used.

File: app/controllers/categories_controller.rb [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+  end
+
+  def destroy
+    @category = Category.find(params[:id])
+    @category.destroy
+
+    respond_to do |format|
+      format.html { redirect_to categories_url, notice: 'Category was successfully destroyed.' }
+      format.json { head :no_content }
+    end
```
Comment (fobadara) [inline app/controllers/categories_controller.rb]: [REQUIRED] What happens when there is an error? Kindly include a fail path for this action.

File: app/controllers/expenses_controller.rb [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+  end
+
+  # DELETE /expenses/1 or /expenses/1.json
+  def destroy
+    @expense.destroy
+
+    respond_to do |format|
+      format.html { redirect_to expenses_url, notice: 'Expense was successfully destroyed.' }
+      format.json { head :no_content }
+    end
```
Comment (fobadara) [inline app/controllers/expenses_controller.rb]: [REQUIRED] Kindly include a fail path that handles errors. This makes your code more professional, saves it from breaking and helps create a good user experience

File: config/routes.rb [hunk 7]
```diff
@@ -1,6 +1,12 @@
...
+  resources :expenses , only: %i[index new create destroy]
```
Comment (fobadara) [inline config/routes.rb:10]: [REQUIRED] I can see an update action in your expenses controller but I cant see the route. Kindly include it or remove the action from the controller.

**Review by fobadara — CHANGES_REQUESTED** [review]
Hi @tjay1760, 

Good job so far!
You still need to work on some issues to go to the next project, but you are almost there!

REDO

## Highlights
✅ Descriptive README
✅ Descriptive PR summary
✅ Clean code
✅ Good first attempt
✅ Good implementation of the principle of separating concerns.
✅ Use of rails partials
✅ All specs are passing

## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you consider them as they can improve your code._

Cheers, and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think the code review was unfair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: README.md [hunk 8]
```diff
@@ -1,24 +1,169 @@
...
+## 👥 Authors <a name="authors"></a>
+
+👤 : **John Murianki**
+
+- GitHub: [@githubhandle](https://github.com/tjay1760)
+- Twitter: [@twitterhandle](https://twitter.com/TjayPod)
+- LinkedIn: [LinkedIn](https://www.linkedin.com/in/john-murianki-thiongo)
+
+
+Original design idea by Gregoire Vella on Behance.
```
Comment (tugsem) [inline README.md]: - [ ]  As requested before, kindly move this line to the acknowledgments section.

File: app/views/categories/index.html.erb (hunk 2 above)
Comment (tugsem) [inline app/views/categories/index.html.erb]: - [ ] As requested, please move this button to the bottom of the page to match the design.
image

File: app/views/categories/index.html.erb [hunk 9]
```diff
@@ -0,0 +1,12 @@
+<%= stylesheet_link_tag 'categories', media: 'all', 'data-turbolinks-track': 'reload' %>
+
+<div class ="categories">
+  <%= link_to "New category", new_category_path, class: "link_to newcat" %>
+  <% if @categories.empty? %>
+      <p>No categories available</p>
+        <% else %>
+    <% @categories.each do |category| %>
+          <%= link_to render(category), category_path(category), class: "link_to cats" %>
```
Comment (tugsem) [inline app/views/categories/index.html.erb]: - [ ] As requested, please increase the size of the image of the category to match the given design.
image

File: config/routes.rb (hunk 4 above)
Comment (tugsem) [inline config/routes.rb]: - [ ] As requested, kindly remove the 'update' action from the categories route as you don't have it.

File: config/routes.rb (hunk 7 above)
Comment (tugsem) [inline config/routes.rb:10]: - [ ] As requested, please add the 'update' action here as you have it in your controller, or remove it.

File: app/controllers/categories_controller.rb (hunk 5 above)
Comment (tugsem) [inline app/controllers/categories_controller.rb]: - [x] As requested, please handle the case of an error in this request(destroy).
https://github.com/tjay1760/Budget-App/pull/1/files/25ed052ce28d3025370ee0db010779aa76358fff#r1366886142

File: app/controllers/expenses_controller.rb (hunk 6 above)
Comment (tugsem) [inline app/controllers/expenses_controller.rb]: - [x] As requested, please handle the error for this action. 
https://github.com/tjay1760/Budget-App/pull/1#discussion_r1366887946

**Review by tugsem — CHANGES_REQUESTED** [review]
## STATUS: REQUIRED CHANGES ♻️
Hello @tjay1760 👋 , 

Great work on making the changes requested by a previous reviewer 👏 
You've done well implementing some of the requested changes, but there are some issues that you still need to work on to go to the next project but you are almost there!

Alt Text

 ## Highlights 🚀
✔️Good job on making the changes

 ## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._


Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**


------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Review by Owusu-Desmond — APPROVED** [review]
Hi @tjay1760 , 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 



## Highlights
📌 No Linters error ✅
📌 Descriptive pull request✅
📌 Good commit messages ✅
📌 Professional README file ✅

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**